examples/mnist.cpp

```diff
@@ -258,25 +272,50 @@ void testMNISTLoadAndTraining() {
     return;
   }
 
-  glow::PlaceholderBindings bindings;
-  // Get input and output placeholders.
-  auto *A = llvm::cast<glow::Placeholder>(
-      EXIT_ON_ERR(loader.getNodeValueByName(inputName)));
-  auto *E = EXIT_ON_ERR(loader.getSingleOutput());
-  bindings.allocate(mod.getPlaceholders());
+  Placeholder *selectedI{nullptr};
+  if ((errPtr =
+           glow::prepareFunctionForTraining(F, inferBindings, selectedI))) {
+    return;
+  }
+
+  inferBindings.allocate(inferMod.getPlaceholders());
+
+  // Load the model a second time for training.
+  // TODO: remove once EE2 is able to compile in different modes.
+  ExecutionEngine2 EET_(executionBackend);
+  auto &trainMod = EET_.getModule();
+  auto *TF = trainMod.createFunction("lenet_mnist_train");
+  glow::Caffe2ModelLoader trainingLoader("lenet_mnist/predict_net.pb",
+                                         "lenet_mnist/init_net.pb", {inputName},
+                                         {inputType}, *TF, &errPtr);
+
+  if (errPtr) {
+    LOG(ERROR) << "Loader failed to load lenet_mnist model for training.";
+    return;
+  }
 
   Placeholder *selected{nullptr};
-  if ((errPtr = glow::prepareFunctionForTraining(F, bindings, selected))) {
+  if ((errPtr =
+           glow::prepareFunctionForTraining(TF, trainingBindings, selected))) {
     return;
   }
 
   const int numIterations = 30;
+  // Get input placeholder.
+  auto *A = llvm::cast<glow::Placeholder>(
+      EXIT_ON_ERR(trainingLoader.getNodeValueByName(inputName)));
```

Comment:
better to request input placeholders and trainingLoader .getInputVarsMapping(), and obtain particular placeholder by name.

Reply:
Do you mean to avoid the casting and EXIT_ON_ERR and use the mapping and CHECK instead ? (like the code below?) 

llvm::StringRef inputName = "input";
auto inputs = loader.getInputVarsMapping();
CHECK(inputs.size()>0 << "msg")
Placeholder *p = inputs[inputName];
CHECK(p << "msg")

Reply:
auto inputs = loader.getInputVarsMapping();
auto it = inputs.find(inputName);
// EXIT_ON_ERR (it == inputs.end()); - pseudo code.  
Placeholder *p = it->second;

Reply:
I agree with the point you made here.
Interestingly, the Caffe2ModelLoader constructor does not load the inputs into this map (another CTOR of this class does , but not this one) - it loads it implicitly via loadConv.
This is really weird and worth looking at, but outside the scope of this porting.
I'm leaving the existing code as is.